src/libclipper/src/selection_policies.cpp

```diff
@@ -1,344 +1,407 @@
-    const std::vector<VersionedModelId>& candidate_models) {
-  UNUSED(candidate_models);
-  return 0;
+VersionedModelId Exp3Policy::select(Exp3State state,
+                                    std::vector<VersionedModelId>& models) {
+  // Helper function for randomly drawing an arm based on its normalized weight
+  auto rand_num = rand() % 1;
+  VersionedModelId selected_model;
+  auto it = models.begin();
+  while (rand_num >= 0) {
+    rand_num -= (1 - eta) * (state.second[*it]["weight"] / state.first) +
+                eta / state.second.size();
+    selected_model = *it;
+    it++;
+  }
+  return selected_model;
 }
 
-std::vector<PredictTask> NewestModelSelectionPolicy::select_predict_tasks(
-    VersionedModelId state, Query query, long query_id) {
-  std::vector<PredictTask> task_vec;
-  // construct the task and put in the vector
-  task_vec.emplace_back(query.input_, state, 1.0, query_id,
-                        query.latency_micros_);
-  return task_vec;
+std::vector<PredictTask> Exp3Policy::select_predict_tasks(Exp3State state,
+                                                          Query query,
+                                                          long query_id) {
+  auto selected_model = select(state, query.candidate_models_);
+  auto task = PredictTask(query.input_, selected_model, 1.0, query_id,
+                          query.latency_micros_);
+  std::vector<PredictTask> tasks{task};
+  return tasks;
 }
 
-Output NewestModelSelectionPolicy::combine_predictions(
-    VersionedModelId state, Query query, std::vector<Output> predictions) {
-  UNUSED(state);
-  UNUSED(query);
-  // just return the first prediction
-  if (predictions.empty()) {
-    return Output{0.0, std::make_pair("none", 0)};
-  } else {
-    return predictions.front();
-  }
+Output Exp3Policy::combine_predictions(Exp3State /*state*/, Query /*query*/,
+                                       std::vector<Output> predictions) {
+  // Don't need to do anything
+  // TODO: what if predictions is empty?
+  return predictions.front();
 }
 
 std::pair<std::vector<PredictTask>, std::vector<FeedbackTask>>
-NewestModelSelectionPolicy::select_feedback_tasks(VersionedModelId state,
-                                                  FeedbackQuery query,
-                                                  long query_id) {
-  UNUSED(state);
-  UNUSED(query);
-  UNUSED(query_id);
-  return std::make_pair(std::vector<PredictTask>(),
-                        std::vector<FeedbackTask>());
-}
-
-VersionedModelId NewestModelSelectionPolicy::process_feedback(
-    VersionedModelId state, Feedback feedback,
-    std::vector<Output> predictions) {
-  UNUSED(feedback);
-  UNUSED(predictions);
-  return state;
+Exp3Policy::select_feedback_tasks(Exp3State state, FeedbackQuery feedback,
+                                  long query_id) {
+  // Predict Task
+  auto selected_model = select(state, feedback.candidate_models_);
+  auto task1 =
```

Comment:
let's not call these variables `task1` and `task2`. How about `predict_tasks` and `feedback_tasks`?